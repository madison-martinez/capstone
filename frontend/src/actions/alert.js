export const alertActionTypes = {
    SUCCESS: 'ALERT_SUCCESS',
    ERROR: 'ALERT_ERROR',
    RESET: 'ALERT_RESET'
};

function success(message) {
    return {
        type: alertActionTypes.SUCCESS,
        message
    }
};

function error(message) {
    return {
        type: alertActionTypes.ERROR,
        message
    }
};

function reset() {
    return {
        type: alertActionTypes.RESET
    }
};


export const alertActions = {
    success,
    error,
    reset
};

