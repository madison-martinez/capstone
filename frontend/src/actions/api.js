import axios from 'axios';

const baseUrl = "https://localhost:5001/api/";

// eslint-disable-next-line import/no-anonymous-default-export
export default {
    product(url = baseUrl + 'Product/') {

        return {
            fetchAll: () => axios.get(url),
            create: newProduct => axios.post(url, newProduct),
            update: (id, updateProduct) => axios.put(url + id, updateProduct),
            delete: id => axios.delete(url + id),
            fetchById: id => axios.get(url + id)
        }
    }
};
