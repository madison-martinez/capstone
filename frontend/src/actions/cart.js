export const CartActionTypes = {
  ADD_TO_CART: "ADD_TO_CART",
  GET_CART_COUNT: "GET_CART_COUNT",
  DELETE_CART_PRODUCT: "DELETE_CART_PRODUCT",
  CLEAR_CART: "CLEAR_CART",
};

export const addToCart = ({
  product,
  id,
  price,
  description,
  quantity,
  image,
}) => {
  const cartProduct = { product, id, price, description, quantity, image };
  return (dispatch) => {
    dispatch({
      type: CartActionTypes.ADD_TO_CART,
      payload: cartProduct,
    });
  };
};

export const getCartCount = () => {
  return (dispatch) => {
    dispatch({
      type: CartActionTypes.GET_CART_COUNT,
    });
  };
};

export const deleteCartProduct = ({ id, price }) => {
  const deletedItem = { id, price };
  return (dispatch) => {
    dispatch({
      type: CartActionTypes.DELETE_CART_PRODUCT,
      payload: deletedItem,
    });
  };
};

export const clearCart = () => {
  return (dispatch) => {
    dispatch({
      type: CartActionTypes.CLEAR_CART,
    });
  };
};
