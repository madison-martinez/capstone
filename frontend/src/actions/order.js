
export const OrderActionTypes = {
  PLACE_ORDER: "PLACE_ORDER",
};

export const placeOrder = ({ products, price, userId }) => {
  const order = { products, price, userId };
  return async (dispatch) => {
    dispatch({
      type: OrderActionTypes.PLACE_ORDER,
      payload: order,
    });
  };
};
