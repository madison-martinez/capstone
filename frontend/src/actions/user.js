import { userService } from "../services/user.service";
import { alertActions } from "./alert";

export const userConstants = {
  REGISTER_REQUEST: "USERS_REGISTER_REQUEST",
  REGISTER_SUCCESS: "USERS_REGISTER_SUCCESS",
  REGISTER_FAILURE: "USERS_REGISTER_FAILURE",

  LOGIN_REQUEST: "USERS_LOGIN_REQUEST",
  LOGIN_SUCCESS: "USERS_LOGIN_SUCCESS",
  LOGIN_FAILURE: "USERS_LOGIN_FAILURE",

  LOGOUT: "USERS_LOGOUT",
};

function login(username, password, handleOnSuccess) {
  return async (dispatch) => {
    dispatch(request({ username }));

    return userService.login(username, password).then(
      (user) => {
        dispatch(success(user));
        handleOnSuccess();
      },
      (error) => {
        dispatch(failure(error.toString()));
        dispatch(alertActions.error(error.toString()));
      }
    );
  };

  function request(user) {
    return { type: userConstants.LOGIN_REQUEST, user };
  }
  function success(user) {
    return { type: userConstants.LOGIN_SUCCESS, user };
  }
  function failure(error) {
    return { type: userConstants.LOGIN_FAILURE, error };
  }
}

function logout() {
  userService.logout();
  return { type: userConstants.LOGOUT };
}

function register(user, handleOnSuccess) {
  return async (dispatch) => {
    dispatch(request(user));

    userService.register(user).then(
      (user) => {
        dispatch(success(user));
        handleOnSuccess();
      },
      (error) => {
        dispatch(failure(error.toString()));
      }
    );
  };

  function request(user) {
    return { type: userConstants.REGISTER_REQUEST, user };
  }
  function success(user) {
    return { type: userConstants.REGISTER_SUCCESS, user };
  }
  function failure(error) {
    return { type: userConstants.REGISTER_FAILURE, error };
  }
}

export const userActions = {
  login,
  logout,
  register,
};
