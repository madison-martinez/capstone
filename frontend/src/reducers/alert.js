import { alertActionTypes } from "../actions/alert";

export function alert(state = {}, action) {
  switch (action.type) {
    case alertActionTypes.SUCCESS:
      return {
        type: "ALERT_SUCCESS",
        message: action.message,
      };
    case alertActionTypes.ERROR:
      return {
        type: "ALERT_ERROR",
        message: action.message,
      };
    case alertActionTypes.RESET:
      return {};
    default:
      return state;
  }
};
