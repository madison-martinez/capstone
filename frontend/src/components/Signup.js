import React, { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import Image from 'next/image';
import Form from './styles/FormStyles';
import useForm from '../utils/useForm';
import { userActions } from '../actions/user';


function Signup() {
  const { values, handleChange, clearForm } = useForm({
    firstName: '',
    lastName: '',
    username: '',
    password: '',
  });

  const [submitted, setSubmitted] = useState(false);
  const { firstName, lastName, username, password } = values;
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(userActions.logout());
  }, []);

  return (
    <Form
      method="post"
      onSubmit={async e => {
        e.preventDefault();
        setSubmitted(true);
        if (firstName && lastName && username && password) {
          const res = await dispatch(userActions.register(values));
          console.log(res);
        }
        clearForm();
      }}
    >
      <fieldset>
        <h2>Sign Up for An Account</h2>

        <label htmlFor="firstName">
          First Name
          <input
            type="text"
            name="firstName"
            placeholder="First Name"
            value={values.firstName}
            onChange={handleChange}
            required
          />
        </label>
        <label htmlFor="lastName">
          Last Name
          <input
            type="text"
            name="lastName"
            placeholder="Last Name"
            value={values.lastName}
            onChange={handleChange}
            required
          />
        </label>
        <label htmlFor="username">
          Username
          <input
            type="text"
            name="username"
            placeholder="Username"
            value={values.username}
            onChange={handleChange}
            required
          />
        </label>
        <label htmlFor="password">
          Password
          <input
            type="password"
            name="password"
            placeholder="Password"
            value={values.password}
            onChange={handleChange}
            required
          />
        </label>

        <button type="submit">Sign Up!</button>
      </fieldset>

      <Image
        src="/assets/photo-1444459094717-a39f1e3e0903.jpeg"
        alt=""
        layout="responsive"
        width={200}
        height={200}
      />

    </Form>
  );
};

export default Signup;