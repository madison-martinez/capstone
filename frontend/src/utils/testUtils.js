import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { reducers } from '../reducers';

function render(
  ui,
  {
    initialState,
    store = createStore(reducers, initialState ),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

//mock middleware functions
const thunk = ({ dispatch, getState }) => next => action => {
    if (typeof action === 'function') {
      return action(dispatch, getState)
    }
    return next(action)
};

const fakeItem = () => ({
    id: '1',
    price: 20.00,
    title: 'Roses',
    description: 'Roses are Red',
});

export {
    fakeItem,
    thunk
};

export * from '@testing-library/react';
export { render };