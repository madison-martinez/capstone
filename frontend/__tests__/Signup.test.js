import thunk from "redux-thunk";
import configureStore from "redux-mock-store";
import { rest } from "msw";
import { setupServer } from "msw/node";
import fetchMock from "fetch-mock";
import userEvent from "@testing-library/user-event";
import { waitFor } from "@testing-library/react";
import { render, screen } from "../src/utils/testUtils";
import { alertActionTypes } from "../src/actions/alert";
import { userConstants, userActions } from "../src/actions/user";
import Signup from "../src/components/Signup";

const middleware = [thunk];
const mockStore = configureStore(middleware);
const initialState = {
  registration: {},
};

const mockSuccessFn = jest.fn();

const mockedUser = {
  firstName: "Pam",
  lastName: "Beasley",
  username: "receptionist01",
  password: "jim",
  role: "",
};

const mockedUserMissingFields = {
  firstName: "Pam",
  lastName: "",
  username: "receptionist01",
  password: "",
  role: "",
};

const server = setupServer(
  rest.post("https://localhost:5001/users/register", (req, res, ctx) => {
    return res(ctx.json(mockedUser));
  })
);

beforeAll(() => server.listen());
afterEach(() => {
  server.resetHandlers();
  jest.clearAllMocks();
  fetchMock.restore();
});
afterAll(() => server.close());

describe("<Signup />", () => {
  let store;

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it("should show an empty signup form when page renders", () => {
    const { container } = render(<Signup />);
    expect(screen.queryByTestId("signup-form")).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it("should allow a user to fill out the form and submit it", async () => {
    render(<Signup store={store} />);
    expect(screen.queryByTestId("signup-form")).toBeTruthy();
    //user fills out form
    userEvent.type(
      screen.getByPlaceholderText("First Name"),
      mockedUser.firstName
    );
    userEvent.type(
      screen.getByPlaceholderText("Last Name"),
      mockedUser.lastName
    );
    userEvent.type(
      screen.getByPlaceholderText("Username"),
      mockedUser.username
    );
    userEvent.type(
      screen.getByPlaceholderText("Password"),
      mockedUser.password
    );

    //user click the submit button
    userEvent.click(screen.getByText("Sign Up!"));

    // expected actions upon submit
    const expectedActions = [
      {
        type: userConstants.REGISTER_REQUEST,
        user: mockedUser,
      },
      {
        type: userConstants.REGISTER_SUCCESS,
        user: mockedUser,
      },
    ];

    return store
      .dispatch(userActions.register(mockedUser, mockSuccessFn))
      .then(async () => {
        const actualActions = store.getActions();
        expect(actualActions).toEqual(expectedActions);
        await waitFor(() => expect(mockSuccessFn).toHaveBeenCalledTimes(1));
      });
  });

  it("should show an error if a required field is not filled out", () => {
    server.use(
      rest.post("https://localhost:5001/users/register", (req, res, ctx) => {
        return res(ctx.status(400));
      })
    );

    render(<Signup store={store} />);
    userEvent.click(screen.getByText("Sign Up!"));

    fetchMock.getOnce("/users/register", {
      body: {},
      headers: { "Content-Type": "application/json" },
    });

    const expectedActions = [
      {
        type: userConstants.REGISTER_REQUEST,
        user: mockedUserMissingFields,
      },
      {
        type: userConstants.REGISTER_FAILURE,
        error: "Bad Request",
      },
      {
        type: alertActionTypes.ERROR,
        message: "Bad Request",
      },
    ];

    return store
      .dispatch(userActions.register(mockedUserMissingFields, mockSuccessFn))
      .then(async () => {
        const actualActions = store.getActions();
        expect(actualActions).toEqual(expectedActions);
      });
  });
});
