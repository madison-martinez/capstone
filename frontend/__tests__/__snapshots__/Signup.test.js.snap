// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> should show an empty signup form when page renders 1`] = `
<div>
  <form
    class="FormStyles__Form-sc-1qtbzh2-0 cmkxQl"
    data-testid="signup-form"
    method="post"
  >
    <fieldset>
      <h2>
        Sign Up for An Account
      </h2>
      <label
        for="firstName"
      >
        First Name
        <input
          name="firstName"
          placeholder="First Name"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        for="lastName"
      >
        Last Name
        <input
          name="lastName"
          placeholder="Last Name"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        for="username"
      >
        Username
        <input
          autocomplete="username"
          name="username"
          placeholder="Username"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password
        <input
          name="password"
          placeholder="Password"
          required=""
          type="password"
          value=""
        />
      </label>
      <label
        for="role"
      >
        Select Your Profession
        <select
          type="select"
        >
          <option
            default=""
          />
          <option
            value="farmer"
          >
            Farmer
          </option>
          <option
            value="restaurant"
          >
            Restaurant
          </option>
        </select>
      </label>
      <button
        type="submit"
      >
        Sign Up!
      </button>
    </fieldset>
    <img
      alt="berries"
      height="650"
      src="/assets/photo-1444459094717-a39f1e3e0903.jpeg"
      width="500"
    />
  </form>
</div>
`;
